### PR TITLE
last bit of charge.py

### PR DESCRIPTION
**Pull request by mganahl**
Add support of BaseCharge for BlockSparseTensors

### CODE REVIEW

File: tensornetwork/block_sparse/charge.py [hunk 1]
```diff
@@ -108,6 +108,17 @@ def charges(self):
     """
     return self.unique_charges[:, self.charge_labels]
 
+  @property
+  def dtype(self):
+    return self.unique_charges.dtype
+
+  @property
+  def degeneracies(self):
+    return np.sum(
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: Break this up, (took me too long to figure out what function the ", 0" was supposed to be for).

File: tensornetwork/block_sparse/charge.py [hunk 2]
```diff
@@ -264,6 +275,169 @@ def intersect(self, other, assume_unique=False, return_indices=False) -> Any:
...
+      BaseCharge: The sorted unique values.
+      np.ndarray: The indices of the first occurrences of the unique values in the
+        original array. Only provided if `return_index` is True.
+      np.ndarray: The indices to reconstruct the original array from the
+        unique array. Only provided if `return_inverse` is True.
+      np.ndarray: The number of times each of the unique values comes up in the
+        original array. Only provided if `return_counts` is True.      
+    """
+    obj = self.__new__(type(self))
+    tmp = np.unique(
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: Is this always going to return a tuple if all of the arguments are False?

File: tensornetwork/block_sparse/charge.py [hunk 3]
```diff
@@ -264,6 +275,169 @@ def intersect(self, other, assume_unique=False, return_indices=False) -> Any:
...
+        out.append(tmp[n])
+
+    if len(out) == 1:
+      return out[0]
+    if len(out) == 2:
+      return out[0], out[1]
+    if len(out) == 3:
+      return out[0], out[1], out[2]
+    if len(out) == 4:
+      return out[0], out[1], out[2], out[3]
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: Can we not just always return out? When would we return `None`?

File: tensornetwork/block_sparse/charge.py [hunk 4]
```diff
@@ -264,6 +275,169 @@ def intersect(self, other, assume_unique=False, return_indices=False) -> Any:
...
+
+    # construct the map to the reduced charges
+    reduced_ind_labels = map_to_reduced[self.charge_labels]
+    reduced_locs = reduced_ind_labels >= 0
+    new_ind_labels = reduced_ind_labels[reduced_locs].astype(np.int16)
+    obj = self.__new__(type(self))
+    obj.__init__(reduced_charges, new_ind_labels, self.charge_types)
+
+    if return_locations:
+      return obj, strides * np.flatnonzero(reduced_locs).astype(np.uint32)
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: Why is this not int16?

File: tensornetwork/block_sparse/charge.py [hunk 5]
```diff
@@ -264,6 +275,169 @@ def intersect(self, other, assume_unique=False, return_indices=False) -> Any:
...
+                self.charge_types, target_charges.charge_types))
+
+      targets = target_charges.unique_charges
+    else:
+      if target_charges.ndim == 1:
+        targets = np.expand_dims(np.unique(target_charges, axis=0), 0)
+      elif target_charges.ndim == 2:
+        targets = np.unique(target_charges, axis=1)
+      else:
+        raise ValueError("targets.ndim has to be 1 or 2, found {}".format(
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: Please add tests that hit all of these raises. Tests on failure cases accounts for a majority of our missing coverage currently

File: tensornetwork/block_sparse/charge.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/block_sparse/charge.py]: fixed

File: tensornetwork/block_sparse/charge.py (hunk 4 above)
Comment (mganahl) [reply tensornetwork/block_sparse/charge.py]: these are index positions within a large data-vector

File: tensornetwork/block_sparse/charge.py (hunk 5 above)
Comment (mganahl) [reply tensornetwork/block_sparse/charge.py]: will do

File: tensornetwork/block_sparse/charge.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/block_sparse/charge.py]: in that case it just returns a BaseCharge object

File: tensornetwork/block_sparse/charge.py (hunk 5 above)
Comment (mganahl) [reply tensornetwork/block_sparse/charge.py]: fixed

**Comment by mganahl** [comment]
should be good to go

**Pull request merged**